docs/design/phases.md

```diff
@@ -0,0 +1,127 @@
+ - a `/etc/kubernetes/tokens.csv` file that contains one kubeadm-generated token
+   - the content should be in the form `${token},${user},${uuid},${group}` where:
+     - `${token}` is the full token `123456.0123456789abcdef`
+     - `${user}` is currently `kubeadm-node-csr`, should be changed
+     - `${uuid}` is a random UUID created with `"k8s.io/apimachinery/pkg/util/uuid".NewUUID()`
+     - `${group}` is currently `kubeadm:kubelet-bootstrap`, might be changed
+
+### Bootstrap the control plane
+
+First, determine if there is an etcd cluster available:
+ - use an external one if external etcd params are given
+ - spin up one locally if external etcd is not be enabled. It should:
+   - be spun up in a Pod
+   - listen on `localhost:2379`
+   - set `PodSpec.SecurityContext.SELinuxOptions.Type=spc_t`
+   - be at a minimum version `3.0.14`
+   - make a volume out to the host on a given host path
+
+Then, spin up the control plane as static pods:
+ - They are located in `/etc/kubernetes/manifests`
+ - The files are stored as `.yaml` files.
+ - `HostNetwork: true` is present on all Pods since there is no network configured yet
+
+Then kubeadm waits for the control plane to come up as static pods
+
+The master node is patched, the master label is added, and the master is tainted.
+
+The proposed master label for `v1.6` is `node-role.kubernetes.io/master=true`.
+Nodes may also register themselves with the label `node-role.kubernetes.io/node=true` by using the `--node-labels` flag.
+
+The taint for proposed beta use is:
+
+```yaml
+apiVersion: v1
+kind: Node
+spec:
+  taints:
+  - key: node-role.kubernetes.io
+    value: master
+    effect: NoSchedule
+```
+
+If self-hosting should be enabled, do a little bit more:
+ - TODO
+
+### Configure the bootstrap tokens
+
+This phase configures the BootstrapSigner to create the public `cluster-info` ConfigMap in the `kube-public` namespace
+ - "Uploads" the `ca.crt` and the address/port the apiserver is listening to the `cluster-info` ConfigMap in the `kubeconfig` key
+ - Adds all tokens given to it as Secrets with the `bootstrap-token-` prefix
+ - Then the `BootstrapSigner` controller JWS-signs content of the `kubeconfig` ConfigMap key with the token(s).
+
+### Addons
+
+ServiceAccounts for `kube-dns` and `kube-proxy` are created in the `kube-system` namespace.
+
+If RBAC is enabled, then ClusterRoleBindings are created for:
+ - the TLS Bootstrap role (`system:node-bootstrapper`), so tokens in the group `kubeadm:kubelet-bootstrap` only can access what's required for TLS Bootstrapping
+ - kube-proxy; the `kube-proxy` ServiceAccount can gets the privileges in the `system:node-proxier` ClusterRole
+ - kube-dns; the `kube-dns` ServiceAccount can gets the privileges in the `system:kube-dns` ClusterRole
+
+In the future, the base layer of configuration for all kubelets in the cluster will be set in this phase.
+
+Then the rest of the addons follow:
+ - apply the CNI Network Provider file if specified
```

Comment:
Please add, make sure CNI plug-in supports the new RBAC mode if RBAC is enabled.